services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+ */
+public class ProtocolVersionAutoDetectionService extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(ProtocolVersionAutoDetectionService.class);
+  private final AtomicBoolean stop = new AtomicBoolean(false);
+  private final Set<String> allClusters;
+  private final Map<String, ProtocolVersionAutoDetectionStats> clusterToStatsMap = new HashMap<>();
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceControllerMultiClusterConfig multiClusterConfigs;
+  final ScheduledExecutorService executor;
+
+  public ProtocolVersionAutoDetectionService(
+      VeniceParentHelixAdmin parentHelixAdmin,
+      VeniceControllerMultiClusterConfig multiClusterConfigs,
+      MetricsRepository metricsRepository,
+      Optional<Map<String, ProtocolVersionAutoDetectionStats>> clusterToStatsMap) {
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+    this.multiClusterConfigs = multiClusterConfigs;
+    this.allClusters = multiClusterConfigs.getClusters();
+    executor = Executors
+        .newScheduledThreadPool(multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionThreadCount());
+
+    if (clusterToStatsMap.isPresent()) {
+      // If the caller has provided a map of stats, use it - for testing purpose
+      this.clusterToStatsMap.putAll(clusterToStatsMap.get());
+    } else {
+      // Otherwise, create a new map of stats for each cluster - default option
+      for (String clusterName: this.allClusters) {
+        this.clusterToStatsMap.put(
+            clusterName,
+            new ProtocolVersionAutoDetectionStats(
+                metricsRepository,
+                "admin_operation_protocol_version_auto_detection_service_" + clusterName));
+      }
+    }
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    executor.scheduleAtFixedRate(
+        new ProtocolVersionDetectionTask(),
+        0,
+        multiClusterConfigs.getAdminOperationProtocolVersionAutoDetectionIntervalMs(),
+        TimeUnit.MILLISECONDS);
+    LOGGER.info("ProtocolVersionAutoDetectionService is started");
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop.set(true);
+    executor.shutdown();
+
+    try {
+      // Wait for the executor to terminate
+      if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+        executor.shutdownNow(); // Force shutdown if not terminated in 60 seconds
+      }
+    } catch (InterruptedException e) {
+      LOGGER.error("Error shutting down ProtocolVersionAutoDetectionService", e);
+      executor.shutdownNow();
+    }
+  }
+
+  private ProtocolVersionAutoDetectionStats getClusterStat(String clusterName) {
+    return clusterToStatsMap.get(clusterName);
+  }
+
+  class ProtocolVersionDetectionTask implements Runnable {
+    @Override
+    public void run() {
+      LOGGER.info("Started running {}", getClass().getSimpleName());
+      while (!stop.get()) {
```

Comment:
It doesn't look right. We shouldn't use a while loop anymore if "ScheduledExecutorService" is running the schedule.
Let's say "ScheduledExecutorService" kicks off a new ProtocolVersionDetectionTask task every 10 minutes, in one hour, we will have 6 of such while loop running; the threads will be depleted soon.

Reply:
You are right. I have updated the code to run one task for one service, we sleep the thread in the middle. Please take another look.

I really appreciate this comment (and other comments). This one is very good catch. Thank you!